lib/writer/checksum-stream.js

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2017 resin.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict'
+
+const stream = require('stream')
+const crypto = require('crypto')
+const CRC32Stream = require('crc32-stream')
+const _ = require('lodash')
+
+/* eslint-disable no-underscore-dangle */
+
+/**
+ * @summary Get a hash stream
+ * @function
+ * @private
+ * @example
+ * var md5Stream = getHashStream('md5')
+ *
+ * @param {String} algorithm - either `crc32` or anything supported by `crypto.Hash`
+ * @returns {Stream.Transform}
+ */
+const getHashStream = (algorithm) => {
+  if (algorithm === 'crc32') {
+    return new CRC32Stream()
+  }
+  return crypto.createHash(algorithm)
+}
+
+/**
+ * @summary Create an instance of ChecksumStream
+ * @name ChecksumStream
+ * @class
+ */
+class ChecksumStream extends stream.Transform {
+  /**
+   * @summary Create an instance of ChecksumStream
+   * @name ChecksumStream
+   * @class
+   * @param {Object} options - options
+   * @param {String[]} options.algorithms - hash algorithms
+   * @example
+   * var checksum = new ChecksumStream({
+   *   algorithms: [ 'crc32', 'md5' ]
+   * })
+   *
+   * checksum.once('checksum', (checksum) => {
+   *   // checksum: {
+   *   //   crc32: 'EF28AF1C',
+   *   //   md5: ''
+   *   // }
+   * })
+   *
+   * fs.createReadStream( 'os-image.img' )
+   *   .pipe( checksum )
+   *   .pipe( fs.createWriteStream( '/dev/rdisk2' ) )
+   *   .once( 'finish', () => { ... })
+   */
+  constructor (options = {}) {
+    super(options)
+    this.results = {}
+    this.algorithms = options.algorithms || []
+    this.hashes = _.map(this.algorithms, (algorithm) => {
+      return this._createHash(algorithm)
+    })
+  }
+
+  /**
+   * @summary Create & pipe to the Hash streams
+   * @private
+   * @param {String[]} algorithm - hash algorithm
+   * @returns {Stream}
+   * @example
+   * const hash = this._createHash(algorithm)
+   */
+  _createHash (algorithm) {
+    let hash = null
```

Comment:
We can avoid the `let` by using `_.attempt`:

```js
const hash = _.attempt(getHashStream, algorithm);
if (_.isError(hash)) {
    ...
}
```

Reply:
Done!